src/express-secure-handlebars.js

```diff
@@ -20,20 +23,123 @@ function ExpressSecureHandlebars(config) {
+ExpressSecureHandlebars.prototype._getFile = function (filePath, options) {
+
+    var file = this._partialsCache.preprocessed[path.relative('.', filePath)];
+    return file ? Promise.resolve(file) : ExpressHandlebars.prototype._getFile.call(this, filePath, options);
+
+};
+
+
+// _getTemplates literally has the core of getTemplates, except that it takes an array of filePaths as input
+/*
+// getTemplates can be rewritten to use _getTemplates in the following way
+ExpressSecureHandlebars.prototype.getTemplates = function (dirPath, options) {
+    options || (options = {});
+    
+    return this._getDir(dirPath, {cache: options.cache}).map(function (filePath) {
+        return path.join(dirPath, filePath);
+    }).then(function(filePaths) {
+        return this._getTemplates(filePaths, options);
+    }.bind(this));
+};
+*/
+ExpressSecureHandlebars.prototype._getTemplates = function (filePaths, options) {
+    options || (options = {});
+
+    var templates = filePaths.map(function (filePath) {
+        return this.getTemplate(filePath, options);
+    }, this);
+
+    return Promise.all(templates).then(function (templates) {
+        return filePaths.reduce(function (hash, filePath, i) {
+            hash[filePath] = templates[i];
+            return hash;
+        }, {});
+    });
+};
+
 
-/* override ExpressHandlebars.render() to expose filePath as compilerOptions */
+// override render() for on-demand partial preprocessing and (pre-)compilation
 ExpressSecureHandlebars.prototype.render = function (filePath, context, options) {
+    options || (options = {});
+
     // expose filePath as processingFile in compilerOptions for secure-handlebars
-    this.compilerOptions || (this.compilerOptions = {});
     this.compilerOptions.processingFile = filePath;
 
-    return expressHandlebars.prototype.render.call(this, filePath, context, options);
+    // ensure getPartials() is called to fetch raw partials
+    return this._exphbsForRawPartials.getPartials(options).then(function(rawPartialsCache) {
+
+        // in constructor, this.compilerOptions.shbsPartialsCache = this._partialsCache
+        this._partialsCache.raw = rawPartialsCache;
+
+        // this._partialsCache.preprocessed is filled by the underlying (pre-)compile() in getTemplate()
+        // TODO: improve this by keeping the returned (pre-)compiled template
+        return this.getTemplate(filePath, options).catch(function(err){
+            throw err;
+        });
+
+    }.bind(this)).then(function() {
+
+        var partialKeys = Object.keys(this._partialsCache.preprocessed);
+
+        // disable preprocessing partials in partial
+        this.compilerOptions.enablePartialProcessing = false;
+        
+        // Merge render-level and (pre-)compiled pre-processed partials together
+        return Promise.all([
+            this._getTemplates(partialKeys, options),   // (pre-)compile pre-processed partials
```

Comment:
`options.partials` supplied to `.render()` must be in (pre-)compiled format, as shown in https://github.com/ericf/express-handlebars/blob/master/lib/express-handlebars.js#L199-L204

`_getTemplates()` here will also produce (pre-)compiled partials